File: leetcode/201-300/286-walls-and-gates/bench/wallsgates.c
// C mirror of wallsgates_seq.kara — see ../README.md § Benchmarks.
// Same LCG, same board parameters, same flat queue, same build-once/punch-many
// shape. Heap-allocated per board and per solve, matching what the Kara, Rust
// and Go lanes do; a stack buffer here would measure a different program.
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define INF 2147483647LL

static long long *make_board(long long b, long long n) {
    long long *g = malloc((size_t)n * sizeof(long long));
    long long s = 777LL + b * 1013LL;
    for (long long i = 0; i < n; i++) {
        s = (s * 1103515245LL + 12345LL) % 2147483648LL;
        long long roll = s % 100LL;
        if (roll < 20) g[i] = -1;
        else if (roll < 21) g[i] = 0;
        else g[i] = INF;
    }
    return g;
}

static long long solve(const long long *template, long long r, long long c,
                       long long *out_unreachable) {
    long long n = r * c;
    long long *g = malloc((size_t)n * sizeof(long long));
    memcpy(g, template, (size_t)n * sizeof(long long));

    long long *q = malloc((size_t)n * sizeof(long long));
    long long tail = 0;
    for (long long k = 0; k < n; k++)
        if (g[k] == 0) q[tail++] = k;

    for (long long head = 0; head < tail; head++) {
        long long cell = q[head];
        long long row = cell / c, col = cell % c;
        long long d = g[cell] + 1;
        long long nb;
        if (row > 0)     { nb = cell - c; if (g[nb] == INF) { g[nb] = d; q[tail++] = nb; } }
        if (row < r - 1) { nb = cell + c; if (g[nb] == INF) { g[nb] = d; q[tail++] = nb; } }
        if (col > 0)     { nb = cell - 1; if (g[nb] == INF) { g[nb] = d; q[tail++] = nb; } }
        if (col < c - 1) { nb = cell + 1; if (g[nb] == INF) { g[nb] = d; q[tail++] = nb; } }
    }

    long long total = 0, unreachable = 0;
    for (long long j = 0; j < n; j++) {
        if (g[j] == INF) unreachable++;
        else if (g[j] > 0) total += g[j];
    }
    free(q);
    free(g);
    *out_unreachable = unreachable;
    return total;
}

typedef struct { long long total, unreachable; } BoardResult;

static BoardResult run_board(long long b, long long r, long long c, long long reps) {
    long long *template = make_board(b, r * c);
    BoardResult out = {0, 0};
    for (long long rep = 0; rep < reps; rep++) {
        long long u = 0;
        out.total += solve(template, r, c, &u);
        out.unreachable += u;
    }
    free(template);
    return out;
}

int main(void) {
    const long long boards = 16, r = 512, c = 512, reps = 8;
    long long total = 0, unreachable = 0;
    for (long long b = 0; b < boards; b++) {
        BoardResult part = run_board(b, r, c, reps);
        total += part.total;
        unreachable += part.unreachable;
    }
    printf("%lld %lld\n", total, unreachable);
    return 0;
}

File: leetcode/201-300/286-walls-and-gates/bench/wallsgates.rs
// Rust mirror of wallsgates_seq.kara — see ../README.md § Benchmarks.
// Same LCG, same board parameters, same flat queue, same build-once/punch-many
// shape. Built twice by bench.sh: once plain (`rustc -O`) and once with
// overflow checks on, which is the equal-safety comparator against Kara's
// always-checked arithmetic.

const INF: i64 = 2147483647;

fn make_board(b: i64, n: i64) -> Vec<i64> {
    let mut g = Vec::with_capacity(n as usize);
    let mut s = 777i64 + b * 1013;
    for _ in 0..n {
        s = (s.wrapping_mul(1103515245) + 12345) % 2147483648;
        let roll = s % 100;
        g.push(if roll < 20 { -1 } else if roll < 21 { 0 } else { INF });
    }
    g
}

fn solve(template: &[i64], r: i64, c: i64) -> (i64, i64) {
    let n = (r * c) as usize;
    let mut g = template.to_vec();

    let mut q: Vec<i64> = Vec::with_capacity(n);
    for k in 0..n {
        if g[k] == 0 {
            q.push(k as i64);
        }
    }

    let mut head = 0usize;
    while head < q.len() {
        let cell = q[head];
        head += 1;
        let row = cell / c;
        let col = cell % c;
        let d = g[cell as usize] + 1;
        let relax = |q: &mut Vec<i64>, g: &mut Vec<i64>, nb: i64| {
            if g[nb as usize] == INF {
                g[nb as usize] = d;
                q.push(nb);
            }
        };
        if row > 0 { relax(&mut q, &mut g, cell - c); }
        if row < r - 1 { relax(&mut q, &mut g, cell + c); }
        if col > 0 { relax(&mut q, &mut g, cell - 1); }
        if col < c - 1 { relax(&mut q, &mut g, cell + 1); }
    }

    let mut total = 0i64;
    let mut unreachable = 0i64;
    for j in 0..n {
        if g[j] == INF { unreachable += 1; } else if g[j] > 0 { total += g[j]; }
    }
    (total, unreachable)
}

fn run_board(b: i64, r: i64, c: i64, reps: i64) -> (i64, i64) {
    let template = make_board(b, r * c);
    let mut total = 0i64;
    let mut unreachable = 0i64;
    for _ in 0..reps {
        let (t, u) = solve(&template, r, c);
        total += t;
        unreachable += u;
    }
    (total, unreachable)
}

fn main() {
    let (boards, r, c, reps) = (16i64, 512i64, 512i64, 8i64);
    let mut total = 0i64;
    let mut unreachable = 0i64;
    for b in 0..boards {
        let (t, u) = run_board(b, r, c, reps);
        total += t;
        unreachable += u;
    }
    println!("{} {}", total, unreachable);
}
